### PR TITLE
Misc fixes for the webhook code

### PR DESCRIPTION
**Pull request by bpradipt**
The webhook code was not updated since a long time resulting in using of old controller-runtime code and annotations which are no longer valid.
Also fixes the tests and some general cleanups of the webhook code for easier readability.

Test results
```
INFO: run tests
webhook_tests.bats
 ✓ [webhook] test it can mutate a pod
 ✓ [webhook] test it should not mutate non-peerpods
 ✓ [webhook] test default parameters can be changed

3 tests, 0 failures

INFO: clean up the test environment
make[1]: Entering directory '/root/cloud-api-adaptor/src/webhook'
kind delete cluster --name "webhook"
Deleting cluster "webhook" ...
Deleted nodes: ["webhook-control-plane"]
make[1]: Leaving directory '/root/cloud-api-adaptor/src/webhook'
Untagged: quay.io/bpradipt/peer-pods-webhook:latest
Deleted: sha256:599b2a22afe9e73ff1350dfe11c4fd88684f5ce628e2d74578720d34baacea62
```

### CODE REVIEW

File: src/webhook/pkg/mutating_webhook/mutating-webhook.go [hunk 1]
```diff
@@ -36,14 +36,15 @@ type PodMutator struct {
 
 // podMutator adds peer-pod extended resource to the pod spec add removes all other resource specs
 func (a *PodMutator) Handle(ctx context.Context, req admission.Request) admission.Response {
+
```
Comment (c3d) [inline src/webhook/pkg/mutating_webhook/mutating-webhook.go:39]: Is that extra blank line intentional?

File: src/webhook/config/manager/manager.yaml [hunk 2]
```diff
@@ -27,41 +27,39 @@ spec:
...
-        # TODO(user): Configure the resources accordingly based on the project requirements.
-        # More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
-        resources:
-          limits:
-            cpu: 500m
-            memory: 128Mi
-          requests:
-            cpu: 10m
-            memory: 64Mi
+        - command:
```
Comment (c3d) [inline src/webhook/config/manager/manager.yaml:30]: All that region is reindented, which obfuscates the actual change, which I believe is only:

```diff
index 097df16..457a16c 100644
--- a/src/webhook/config/manager/manager.yaml
+++ b/src/webhook/config/manager/manager.yaml
@@ -36,8 +36,6 @@ spec:
           env:
             - name: TARGET_RUNTIMECLASS
               value: kata-remote
-        - name: POD_VM_INSTANCE_TYPE
-          value: t2.small
             - name: POD_VM_EXTENDED_RESOURCE
               value: kata.peerpods.io/vm
           securityContext:
```

File: src/webhook/tests/e2e/webhook_tests.bats [hunk 3]
```diff
@@ -23,20 +23,20 @@ assert_pod_mutated() {
 	local expect_vm_limits="$2"
 	local expect_vm_requests="$3"
 
-        local actual_instance_type=$(kubectl get -f "$pod_file" \
```
Comment (c3d) [inline src/webhook/tests/e2e/webhook_tests.bats:26]: Also whitespace changes that obfuscate the actual change a bit:

```diff
index 8831491..fcd3410 100644
--- a/src/webhook/tests/e2e/webhook_tests.bats
+++ b/src/webhook/tests/e2e/webhook_tests.bats
@@ -60,10 +60,10 @@ teardown() {
 
 @test "$test_tags test it should not mutate non-peerpods" {
        echo "Create a pod without runtimeClassName"
-       cat "$pod_file" | sed -e 's/^\s*runtimeClassName:.*//' | \
+       cat "$pod_file" | sed -e 's/^\s*runtimeClassName:.*//' |
                kubectl apply -f -
 
-       ! kubectl get -f ../../hack/pod.yaml -o json | \
+       ! kubectl get -f ../../hack/pod.yaml -o json |
                grep kata\.peerpods
 }
 
@@ -88,10 +88,7 @@ teardown() {
        kubectl set env deployment/peer-pods-webhook-controller-manager \
                -n peer-pods-webhook-system TARGET_RUNTIMECLASS="$runtimeclass"
 
-       kubectl set env deployment/peer-pods-webhook-controller-manager \
-               -n peer-pods-webhook-system POD_VM_INSTANCE_TYPE="$instance_type"
-
-       cat "$pod_file" | sed -e 's/^\(\s*runtimeClassName:\).*/\1 '${runtimeclass}'/' | \
+       cat "$pod_file" | sed -e 's/^\(\s*runtimeClassName:\).*/\1 '${runtimeclass}'/' |
                kubectl apply -f -
 
        kubectl get -f $pod_file -o json
```

File: src/webhook/tests/e2e/webhook_tests.bats [hunk 4]
```diff
@@ -60,10 +60,10 @@ teardown() {
 
 @test "$test_tags test it should not mutate non-peerpods" {
 	echo "Create a pod without runtimeClassName"
-	cat "$pod_file" | sed -e 's/^\s*runtimeClassName:.*//' | \
```
Comment (c3d) [inline src/webhook/tests/e2e/webhook_tests.bats:63]: Why did you remove the trailing `\` ?

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 5]
```diff
@@ -1,17 +1,26 @@
 package mutating_webhook
 
 import (
+	"log"
```
Comment (c3d) [inline src/webhook/pkg/mutating_webhook/remove-resourcespec.go:4]: Nit: extra space in cover letter between "Add" and "peer-pod resource"

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 6]
```diff
@@ -1,17 +1,26 @@
...
 	"k8s.io/apimachinery/pkg/api/resource"
 )
 
 const (
 	RUNTIME_CLASS_NAME_DEFAULT       = "kata-remote"
 	POD_VM_EXTENDED_RESOURCE_DEFAULT = "kata.peerpods.io/vm"
+	PEERPODS_CPU_ANNOTATION          = "io.katacontainers.config.hypervisor.default_vcpus"
+	PEERPODS_MEMORY_ANNOTATION       = "io.katacontainers.config.hypervisor.default_memory"
+	GPU_RESOURCE_NAME                = "nvidia.com/gpu"
+	PEERPODS_GPU_ANNOTATION          = "io.katacontainers.config.hypervisor.default_gpus"
```
Comment (c3d) [inline src/webhook/pkg/mutating_webhook/remove-resourcespec.go:19]: Where is this annotation implemented? I can't find it in the runtime (either main or CCv0)

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 7]
```diff
@@ -26,18 +35,98 @@ func (a *PodMutator) mutatePod(pod *corev1.Pod) (*corev1.Pod, error) {
...
+	memoryLimit := utils.GetResourceLimitQuantity(pod, corev1.ResourceMemory)
+
+	// Get total GPU resource requests
+	gpuRequest := utils.GetResourceRequestQuantity(pod, corev1.ResourceName(GPU_RESOURCE_NAME))
+
+	// log the resource values
+	logger.Printf("CPU Request: %s, CPU Limit: %s, Memory Request: %s, Memory Limit: %s, GPU Request: %s",
+		cpuRequest.String(), cpuLimit.String(), memoryRequest.String(), memoryLimit.String(), gpuRequest.String())
+
+	logger.Printf("Memory: %s", memoryRequest.String())
```
Comment (stevenhorsman) [inline src/webhook/pkg/mutating_webhook/remove-resourcespec.go:68]: Do we need this log statement, when it is in the one above, or have I missed something?

File: src/webhook/pkg/utils/utils.go [hunk 8]
```diff
@@ -58,62 +54,40 @@ func GetResourceRequest(pod *corev1.Pod, resource corev1.ResourceName) int64 {
...
 
-	for i := range pod.Spec.Containers {
-		annotations = mergeContainerResources(&pod.Spec.Containers[i], defaultRequirements, "container", annotations)
+	switch resourceName {
+	case corev1.ResourceCPU:
+		limitQuantity = resource.Quantity{Format: resource.DecimalSI}
+	case corev1.ResourceMemory:
+		limitQuantity = resource.Quantity{Format: resource.BinarySI}
+	default:
+		limitQuantity = resource.Quantity{Format: resource.DecimalSI}
```
Comment (stevenhorsman) [inline src/webhook/pkg/utils/utils.go:67]: Optional discussion point. What do you think about refactoring this out to something like `getResourceQuantity(resourceName)` so that it can be shared by `GetResourceLimitQuantity` and `GetResourceRequestQuantity`?

I was trying to think if we could share more of the code between them, but I'm not sure we should go down the reflection route!

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 9]
```diff
@@ -26,18 +35,98 @@ func (a *PodMutator) mutatePod(pod *corev1.Pod) (*corev1.Pod, error) {
...
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// A non-existent request or limit will be 0 and we don't need to add annotation for 0
+	// We only add annotation if the value is greater than 0
+	// Limit will always be preferred over request
+
+	// Add cpu annotation
+	if cpuRequest.Cmp(resource.MustParse("0")) != 0 && cpuLimit.Cmp(cpuRequest) >= 0 {
```
Comment (stevenhorsman) [inline src/webhook/pkg/mutating_webhook/remove-resourcespec.go:84]: Could these be made more readable with https://github.com/kubernetes/apimachinery/blob/v0.30.2/pkg/api/resource/quantity.go#L520?
e.g. 
```
if !cpuRequest.isZero() && cpuLimit.Cmp(cpuRequest) >= 0 {
```
...
Assume that has the same effect as the current code?
I appreciate it's not much, but it took my tired brain a while to process `Cmp(resource.MustParse("0")) != 0`!

File: src/webhook/pkg/mutating_webhook/mutating-webhook_test.go [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+
+	// Check that the first container has the peer-pod resource added
+	expectedResource := resource.MustParse("1")
+	if !mutatedPod.Spec.Containers[0].Resources.Requests[corev1.ResourceName(POD_VM_EXTENDED_RESOURCE_DEFAULT)].Equal(expectedResource) {
+		t.Errorf("Expected VM resource request to be 1, got %v",
+			mutatedPod.Spec.Containers[0].Resources.Requests[corev1.ResourceName(POD_VM_EXTENDED_RESOURCE_DEFAULT)])
+	}
+}
+
+func TestMutatePod_Negative(t *testing.T) {
```
Comment (stevenhorsman) [inline src/webhook/pkg/mutating_webhook/mutating-webhook_test.go:514]: Optional suggestion: Maybe the name `TestMutatePod_NoChangeForDifferentRuntimeClass` would be easier to understand at a glance in test output?

File: src/webhook/config/manager/manager.yaml (hunk 2 above)
Comment (bpradipt) [reply src/webhook/config/manager/manager.yaml:30]: Sorry about it.. I'll redo these changes. I didn't realise my IDE auto-formatted it

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go (hunk 7 above)
Comment (bpradipt) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:68]: No.... It's stale

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go (hunk 9 above)
Comment (bpradipt) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:84]: Good suggestion :-) ..

File: src/webhook/pkg/utils/utils.go (hunk 8 above)
Comment (bpradipt) [reply src/webhook/pkg/utils/utils.go:67]: I can refactor this as well.

File: src/webhook/tests/e2e/webhook_tests.bats (hunk 4 above)
Comment (bpradipt) [reply src/webhook/tests/e2e/webhook_tests.bats:63]: Right. It's not needed due to `I`

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go (hunk 6 above)
Comment (bpradipt) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:19]: This is hallucination :-( ..
I wanted to just add a custom annotation to indicate the gpu resources in the original request for potential use in cloud-api-adaptor code later on. Forgot to add the right commit.

File: src/webhook/pkg/mutating_webhook/mutating-webhook.go (hunk 1 above)
Comment (bpradipt) [reply src/webhook/pkg/mutating_webhook/mutating-webhook.go:39]: No. Will fix it in next iteration

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 11]
```diff
@@ -26,18 +35,96 @@ func (a *PodMutator) mutatePod(pod *corev1.Pod) (*corev1.Pod, error) {
...
+		logger.Printf("Adding Memory annotation based on memoryRequest: %s", memoryRequest.String())
+		memoryRequestMiBStr, err := utils.ConvertMemoryQuantityToMib(memoryRequest)
+		if err != nil {
+			logger.Printf("Error converting memory quantity to MiB: %v", err)
+		}
+		annotations[PEERPODS_MEMORY_ANNOTATION] = memoryRequestMiBStr
+	}
+
+	// Add GPU annotation
+	if gpuRequest.Cmp(resource.MustParse("0")) > 0 {
```
Comment (stevenhorsman) [inline src/webhook/pkg/mutating_webhook/remove-resourcespec.go:109]: And here?

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 12]
```diff
@@ -26,18 +35,96 @@ func (a *PodMutator) mutatePod(pod *corev1.Pod) (*corev1.Pod, error) {
...
+
+	// A non-existent request or limit will be 0 and we don't need to add annotation for 0
+	// We only add annotation if the value is greater than 0
+	// Limit will always be preferred over request
+
+	// Add cpu annotation
+	if !cpuRequest.IsZero() && cpuLimit.Cmp(cpuRequest) >= 0 {
+		logger.Printf("Adding CPU annotation based on cpuLimit: %s", cpuLimit.String())
+		annotations[PEERPODS_CPU_ANNOTATION] = cpuLimit.String()
+	} else if cpuRequest.Cmp(resource.MustParse("0")) > 0 {
```
[the remark this answers is not in the document]
Comment (bpradipt) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:85]: I want the quantity to be positive. May be Sign can be used
```
// Sign returns 0 if the quantity is zero, -1 if the quantity is less than zero, or 1 if the
// quantity is greater than zero.
func (q *Quantity) Sign() int {
	if q.d.Dec != nil {
		return q.d.Dec.Sign()
	}
	return q.i.Sign()
}
```

[the remark this answers is not in the document]
Comment (stevenhorsman) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:85]: ah sorry, I missed that distinction, `Sign` sounds okay, but equally if you want to leave it as is then that's fine.

**Comment by bpradipt** [comment]
@stevenhorsman any idea on why the test is failing. I think `TestCloudServiceWithSecureComms` is failing

**Comment by stevenhorsman** [comment]
> @stevenhorsman any idea on why the test is failing. I think `TestCloudServiceWithSecureComms` is failing

That test is a bit flakey, which we probably need to ask David to look into, but I've re-run it, so hopefully it will work this time.

**Review by stevenhorsman — APPROVED** [review]
LGTM. Thanks for the updates

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 13]
```diff
@@ -26,18 +35,96 @@ func (a *PodMutator) mutatePod(pod *corev1.Pod) (*corev1.Pod, error) {
...
+	}
+
+	// Add memory annotation
+	if !memoryRequest.IsZero() && memoryLimit.Cmp(memoryRequest) >= 0 {
+		logger.Printf("Adding Memory annotation based on memoryLimit: %s", memoryLimit.String())
+		memoryLimitMiBStr, err := utils.ConvertMemoryQuantityToMib(memoryLimit)
+		if err != nil {
+			logger.Printf("Error converting memory quantity to MiB: %v", err)
+		}
+		annotations[PEERPODS_MEMORY_ANNOTATION] = memoryLimitMiBStr
```
Comment (wainersm) [inline src/webhook/pkg/mutating_webhook/remove-resourcespec.go:97]: Hi @bpradipt !

Should it really set annotations[PEERPODS_MEMORY_ANNOTATION] if conversion to memoryLimitMiBStr failed?

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go [hunk 14]
```diff
@@ -26,18 +35,96 @@ func (a *PodMutator) mutatePod(pod *corev1.Pod) (*corev1.Pod, error) {
...
+
+	} else if memoryRequest.Sign() == 1 {
+		logger.Printf("Adding Memory annotation based on memoryRequest: %s", memoryRequest.String())
+		memoryRequestMiBStr, err := utils.ConvertMemoryQuantityToMib(memoryRequest)
+		if err != nil {
+			logger.Printf("Error converting memory quantity to MiB: %v", err)
+		}
+		annotations[PEERPODS_MEMORY_ANNOTATION] = memoryRequestMiBStr
```
Comment (wainersm) [inline src/webhook/pkg/mutating_webhook/remove-resourcespec.go:105]: Dito.

File: src/webhook/pkg/mutating_webhook/remove-resourcespec.go (hunk 13 above)
Comment (bpradipt) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:97]: Yeah, it'll be 0 and the error is logged to see why it failed. It's not a strict failure..

Comment (bpradipt) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:97]: Let me know if you want me to add a comment to make it explicit. Although looking at the `ConvertMemoryQuantityToMib` method will make it clear that it returns "0" and the same is set as the annotation.

Comment (wainersm) [reply src/webhook/pkg/mutating_webhook/remove-resourcespec.go:97]: ok. I didn't look at `ConvertMemoryQuantityToMib` and I was worried it returning some invalid value. It's okay as is

**Review by wainersm — APPROVED** [review]
Thanks @bpradipt !

**Review by huoqifeng — APPROVED** [review]
LGTM
